### PR TITLE
Fix value sending in async scenarios

### LINKED ISSUE #974: Value charged multiple times in a cascade of async calls
If we have a cascade of contracts calls:
```
User                  Contract A                                            Contract B
  | -- msg_with_value --> |                                                     |
                          |---- msg::send_bytes_and_wait_for_reply(payload) --->|
                                                                                |
                          |<------------------ reply(output) -------------------|
                          |
  |<------- reply --------|
```
then the value passed to the `Contract A` will be attempted to be transferred every time the contract gets to execute a message (on the first entry, upon being woken up from the wait list etc.) - a journal note `SendValue` is generated and processed by the journal handler.
As a result, if there is enough value reserved for the original message (to pay for gas, not value) then this will go unnoticed except that the user will be overcharged. In the worst case we can fall through into an `unreachable!` branch and panic.

To reproduce:
1. Pull a custom branch https://github.com/gear-tech/gear/blob/ek-expose-bug/pallets/gear/src/tests.rs#L3149.
2. Run the test `cascading_messages_with_value_do_not_overcharge`.

### PR DESCRIPTION
**Pull request by breathx**
Resolves #974.

See `core-processor/src/processor.rs`

Thanks to @ekovalev for bug revealing and providing test for it.

- Bug fixed.
- Docs provided.
- Test updated (previous values asserted with inversion).

@gear-tech/dev

### CODE REVIEW

File: pallets/gear/src/tests.rs [hunk 1]
```diff
@@ -3144,6 +3145,77 @@ fn test_reply_to_terminated_program() {
...
+        // The sender's account has enough funds for both gas and `value`, therefore
+        // expecting the message to be processed successfully.
+        // Expected outcome: the sender's balance has decreased by the (gas + `value`),
+        // that is by 2,302,360,260.
+        //
+        // TODO: instead, after the proxy message has re-entered having been woken up
+        // it attempts to do value transfer again.
+        // The `repatriate_reserved()` in this case still succeeds even though the actual
+        // amount repatriated is less than requested 10_000_000 => must thrown here!!!
+        // However, the sender's balance ends up overcharged.
```
Comment (SabaunT) [inline pallets/gear/src/tests.rs:3188]: seems that behavior is fixed? Maybe should state that and remove **TODO**

File: pallets/gear/src/tests.rs [hunk 2]
```diff
@@ -3144,6 +3145,77 @@ fn test_reply_to_terminated_program() {
...
+
+        assert_ok!(Gear::send_message(
+            Origin::signed(USER_1),
+            wrapper_id,
+            100_u64.to_le_bytes().to_vec(),
+            gas_reserved,
+            value,
+        ));
+
+        run_to_block(3, None);
```
Comment (SabaunT) [inline pallets/gear/src/tests.rs:3203]: shouldn't balance checks be performed before and after that call?

**Review by SabaunT — APPROVED** [review]
nice catch @ekovalev!
nice job @breathx

**Comment by clearloop** [comment]
Anybody run `cargo test` under `pallets/gear` after this commit? got `cascading_messages_with_value_do_not_overcharge` failed on my machine:

```
failures:

---- tests::cascading_messages_with_value_do_not_overcharge stdout ----
thread 'tests::cascading_messages_with_value_do_not_overcharge' panicked at 'assertion failed: `(left == right)`
  left: `497354549734`,
 right: `497336081914`', pallets/gear/src/tests.rs:3205:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Pull request merged**